File: AtriusSamples/AtriusSamples/SampleVCs/ExternalLocationServices.m
//
//  ExternalLocationServices.m

//
//  Copyright © 2021 Atrius, part of Acuity Brands. All rights reserved.
//

#import "ExternalLocationServices.h"
#import <LocusLabsSDK/LocusLabsSDK.h>
#import <LocusLabsSDK/LLConstants.h>

@interface ExternalLocationServices () <LLMapViewDelegate, LLVenueDatabaseDelegate>

@property (nonatomic, weak)   LLMapView         *mapView;
@property (nonatomic, strong) LLVenue           *venue;
@property (nonatomic, strong) LLVenueDatabase   *venueDatabase;

@end

@implementation ExternalLocationServices


#pragma mark Lifecycle

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    [LLLocusLabs setup].accountId = @"A11F4Y6SZRXH4X";

    // Create a new LLMapView, register as its delegate and add it as a subview
    LLMapView *mapView = [[LLMapView alloc] initWithFrame:self.view.bounds];
    mapView.autoresizingMask = (UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight);
    mapView.delegate = self;
    [self.view addSubview:mapView];
    
    self.mapView = mapView;
    self.mapView.showTopSafeAreaOverlay = NO;
    self.mapView.darkMode = self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark;
    
    // Get an instance of LLVenueDatabase, set it's mapview and register as its delegate
    self.venueDatabase = [LLVenueDatabase venueDatabaseWithMapView:self.mapView];
    self.venueDatabase.delegate = self;
    
    [self.venueDatabase loadVenueAndMap:@"lax" block:^(LLVenue *venue, LLMap *map, LLFloor *floor, LLMarker *marker) {
        
        self.venue = venue;
        self.mapView.positionManager.activePositioning = NO;
        self.mapView.positionManager.passivePositioning = NO;
        
        // Set the navigation source to external
        [[NSNotificationCenter defaultCenter] postNotificationName:NOTIFICATION_SET_POSITIONING_SENSOR_ALGORITHM
                                                            object:nil
                                                          userInfo:@{@"positioningSensorAlgorithm":@(LLPositioningSensorAlgorithmExternal)}];
    }];
}

- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    
    [super traitCollectionDidChange:previousTraitCollection];
    
    self.mapView.darkMode = self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark;
}

#pragma mark Custom

// This is a pseudomethod to demonstrate how data may be received from an external provider and how it can be passed to our SDK
- (void)didReceiveExternalLocationDict:(NSDictionary *)externalLocationDict {

    NSString *locusLabsFloorID = [self locusLabsFloorIDForExternalFloorID:externalLocationDict[@"FloorId"]];
    if (!locusLabsFloorID) return;
    
    NSDictionary *locationDict = [self locationDictWithLat:externalLocationDict[@"Lat"] lon:externalLocationDict[@"Lon"] floorId:locusLabsFloorID heading:nil];
    [self postUserPosition:locationDict];
}

- (NSDictionary *)locationDictWithLat:(NSNumber *)lat lon:(NSNumber *)lon floorId:(NSString *)floorId heading:(NSNumber *)heading {

    if (!heading) heading = @(180);
    
    LLLatLng *latLng = [[LLLatLng alloc] initWithLat:lat lng:lon];
    NSDictionary *locationDict = @{@"latLng": latLng,
                                   @"errorRadius": @(1),
                                   @"floorId": floorId,
                                   @"heading": heading}; // heading is optional
    
    return locationDict;
}

- (void)mockExternalLocationData {
    
    // Position 1 (Initial - DFS Duty Free)
    NSDictionary *locationDict = @{@"FloorId": @"T48L3", @"Lat": @33.941485, @"Lon": @-118.40195};
    [self performSelector:@selector(didReceiveExternalLocationDict:) withObject:locationDict afterDelay:1.0];
    
    // Position 2 (2 secs later)
    locationDict = @{@"FloorId": @"T48L3", @"Lat": @33.941398, @"Lon": @-118.401916};
    [self performSelector:@selector(didReceiveExternalLocationDict:) withObject:locationDict afterDelay:3.0];
    
    // Position 3 (4 secs later)
    locationDict = @{@"FloorId": @"T48L3", @"Lat": @33.941283, @"Lon": @-118.401863};
    [self performSelector:@selector(didReceiveExternalLocationDict:) withObject:locationDict afterDelay:5.0];

    // Position 4 (6 secs later)
    locationDict = @{@"FloorId": @"T48L3", @"Lat": @33.941102, @"Lon": @-118.401902};
    [self performSelector:@selector(didReceiveExternalLocationDict:) withObject:locationDict afterDelay:7.0];

    // Position 5 (8 secs later - Destination - Gate 64B)
    locationDict = @{@"FloorId": @"T48L3", @"Lat": @33.940908, @"Lon": @-118.40177};
    [self performSelector:@selector(didReceiveExternalLocationDict:) withObject:locationDict afterDelay:9.0];
}

- (void)postUserPosition:(NSDictionary *)locationDict {
    
    [[NSNotificationCenter defaultCenter] postNotificationName:NOTIFICATION_POSITION_SENSOR_POSITION_CHANGED
                                                        object:nil
                                                      userInfo:locationDict];
}

- (void)hideBlueDot {
    
    [[NSNotificationCenter defaultCenter] postNotificationName:NOTIFICATION_POSITION_CLEAR object:nil];
}

- (NSString *)locusLabsFloorIDForExternalFloorID:(NSString *)externalFloorId {
    
    NSString *locusLabsFloorId = nil;
    if ([externalFloorId isEqualToString:@"T48L3"]) locusLabsFloorId = @"lax-terminal6-departures";
    else if ([externalFloorId isEqualToString:@"???"]) locusLabsFloorId = @"???";
    else if ([externalFloorId isEqualToString:@"???"]) locusLabsFloorId = @"???";
        
    return locusLabsFloorId;
}

#pragma mark Delegates - LLVenueDatabase

- (void)venueDatabase:(LLVenueDatabase *)venueDatabase venueLoadFailed:(NSString *)venueId code:(LLDownloaderError)errorCode message:(NSString *)message {

    // Handle failures here
}

#pragma mark Delegates - LLMapView

- (void)mapViewDidClickBack:(LLMapView *)mapView {
    
    // The user tapped the "Cancel" button while the map was loading. Dismiss the app or take other appropriate action here
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)mapViewReady:(LLMapView *)mapView {
    
    // The map is ready to be used in calls e.g. zooming, showing poi, etc.
    [self mockExternalLocationData];
}

- (void)mapView:(LLMapView *)mapView didTapAtPosition:(LLPosition *)position {
    
    NSLog(@"floor:%@", position.floorId);
}

- (UIViewController *)presentingControllerForMapView:(LLMapView *)mapView forContext:(LLMapViewPresentationContext)context {
    
    // Return a viewcontroller the SDK can use to present alerts
    return self;
}

@end

File: LocusLabsSamplesSwift/LocusLabsSamplesSwift/Samples/ExternalLocationServices.swift
//

//  Copyright © 2020 LocusLabs. All rights reserved.
//

import UIKit

// NB - You may need to run "pod update" before running this code

class ExternalLocationServices: UIViewController, LLVenueDatabaseDelegate, LLMapViewDelegate {

    // Vars
    var venueDatabase:      LLVenueDatabase!
    var venue:              LLVenue?
    var mapView:            LLMapView?
    
    // MARK: Lifecycle
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        // Initialize the LocusLabs SDK with the accountId provided by LocusLabs
        LLLocusLabs.setup().accountId = "A11F4Y6SZRXH4X"
        
        // Create a new LLMapView, register as its delegate and add it as a subview
        mapView = LLMapView(frame: view.bounds)
        mapView?.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView?.delegate = self
        view.addSubview(mapView!)
        
        mapView?.showTopSafeAreaOverlay = false
        mapView?.darkMode = traitCollection.userInterfaceStyle == .dark
        
        // Get an instance of LLVenueDatabase, register as its delegate and load the venue LAX
        venueDatabase = LLVenueDatabase(mapView: mapView)
        venueDatabase.delegate = self
        
        // Load the venue LAX async
        venueDatabase.loadVenueAndMap("lax") { (_venue: LLVenue?, _map: LLMap?, _floor: LLFloor?, _marker: LLMarker?) in
            
            self.venue = _venue
            
            // Set the navigation source to external
            let algorithm = NSNumber(value: Int8(LLPositioningSensorAlgorithmExternal.rawValue))
            NotificationCenter.default.post(name: NSNotification.Name(rawValue: NOTIFICATION_SET_POSITIONING_SENSOR_ALGORITHM), object: nil, userInfo: ["positioningSensorAlgorithm": algorithm])
        }
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        
        super.traitCollectionDidChange(previousTraitCollection)
        
        mapView?.darkMode = traitCollection.userInterfaceStyle == .dark
    }
    
    // MARK: Custom
    // This is a pseudo method to demonstrate how data may be received from an external provider and how it can be passed to our SDK
    func didReceiveExternalLocation(_ dict: [String: Any]) {
        
        if let externalFloorId = dict["FloorId"] as? String {
        
            if let locusLabsFloorId = locusLabsFloorIDFor(externalFloorId: externalFloorId) {
                
                let theLocationDict = locationDict(lat: dict["Lat"] as? NSNumber, lon: dict["Lon"] as? NSNumber, floorId: locusLabsFloorId, heading: nil)
                postUserPosition(locationDict: theLocationDict)
            }
        }
    }
    
    func hideBlueDot() {
        
        NotificationCenter.default.post(name: NSNotification.Name(rawValue: NOTIFICATION_POSITION_CLEAR), object: nil, userInfo: nil)
    }
    
    func locationDict(lat: NSNumber!, lon:NSNumber!, floorId:String!, heading:NSNumber?) -> [String: Any]  {
    
        let latLng = LLLatLng(lat: lat, lng: lon)
        
        var locationDict = [String: Any]()
        locationDict["latLng"] = latLng
        locationDict["errorRadius"] = NSNumber(value: 1)
        locationDict["floorId"] = floorId
        if heading != nil {locationDict["heading"] = heading} // Optional
        
        return locationDict
    }
    
    func locusLabsFloorIDFor(externalFloorId: String!) -> String? {
        
        // If you are not able to compile this mapping table yourself, please send us a list (help@locuslabs.com) of
        // the building names and associated floor ids as provided by your external mapping provider and we will compile the mapping table
        // To obtain a list of all LocusLabs buildings and associated floors for this venue, uncomment the line [self.venue listBuildings] in viewDidLoad
        var locusLabsFloorId: String?
        
        if externalFloorId == "T48L3" {locusLabsFloorId = "lax-south-departures"}
        else if externalFloorId == "???" {locusLabsFloorId = "???"}
        else if externalFloorId == "???" {locusLabsFloorId = "???"}
        
        return locusLabsFloorId
    }
    
    func mockExternalLocationData() {
        
        // Position 1 (Initial - DFS Duty Free)
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) {
            
            let locationDict = ["FloorId":"T48L3", "Lat":NSNumber(value:33.941485), "Lon":NSNumber(value:-118.40195)] as [String : Any]
            self.didReceiveExternalLocation(locationDict)
        }
        
        // Position 2 (2 secs later)
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(3)) {
            
            let locationDict = ["FloorId":"T48L3", "Lat":NSNumber(value:33.941398), "Lon":NSNumber(value:-118.401916)] as [String : Any]
            self.didReceiveExternalLocation(locationDict)
        }
        
        // Position 3 (4 secs later)
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(5)) {
            
            let locationDict = ["FloorId":"T48L3", "Lat":NSNumber(value:33.941283), "Lon":NSNumber(value:-118.401863)] as [String : Any]
            self.didReceiveExternalLocation(locationDict)
        }
        
        // Position 4 (6 secs later)
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(7)) {
            
            let locationDict = ["FloorId":"T48L3", "Lat":NSNumber(value:33.941102), "Lon":NSNumber(value:-118.401902)] as [String : Any]
            self.didReceiveExternalLocation(locationDict)
        }
        
        // Position 5 (8 secs later - Destination - Gate 64B)
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(9)) {
            
            let locationDict = ["FloorId":"T48L3", "Lat":NSNumber(value:33.940908), "Lon":NSNumber(value:-118.40177)] as [String : Any]
            self.didReceiveExternalLocation(locationDict)
        }
    }
    
    func postUserPosition(locationDict: [String: Any]!) {
        
        NotificationCenter.default.post(name: NSNotification.Name(rawValue: NOTIFICATION_POSITION_SENSOR_POSITION_CHANGED), object: nil, userInfo: locationDict)
    }
    
    // MARK: Delegates - LLVenueDatabase
    func venueDatabase(_ venueDatabase: LLVenueDatabase!, venueLoadFailed venueId: String!, code errorCode: LLDownloaderError, message: String!) {
        
        // Handle failures here
    }
    
    // MARK: Delegates - LLMapView
    func mapViewDidClickBack(_ mapView: LLMapView!) {
        
        // The user tapped the "Cancel" button while the map was loading. Dismiss the app or take other appropriate action here
        navigationController?.popViewController(animated: true)
    }
    
    func mapViewReady(_ mapView: LLMapView!) {
        
        // The map is ready to be used in calls e.g. zooming, showing poi, etc.
        mockExternalLocationData()
    }
    
    func presentingController(for mapView: LLMapView!, for context: LLMapViewPresentationContext) -> UIViewController! {
        
        // Return a viewcontroller the SDK can use to present alerts
        return self
    }
}

